### PR TITLE
Add Agent Compose package and runtime interface skeleton

### PR DESCRIPTION
**Pull request by ISEEKYAN**
## Summary

This PR evolves `experimental/agent_compose` from a placeholder into the initial reviewable Agent Compose framework.

It:

- establishes the three-layer `runtime`, `model`, and `primitive` package structure;
- exposes the public Python namespace `megatron.experimental.agent_compose`;
- exposes the backend-neutral Runtime interface and its pretrain, RL-ready, and RL-best tiers;
- exposes shared runtime configuration, batch/result, model-handle, and loss contracts;
- adds the initial agent-agnostic skills constitution and one contract for each architecture layer;
- documents the architecture, Runtime interface, model protocol boundary, and incubation lifecycle;
- adds import, layering, Runtime API, and skill-registry contract tests.

## Main And Dev Relationship

`experimental/agent_compose` on `main` is the independently reviewed upstream incubation surface. The complete work-in-progress implementation remains under `experimental/lite` on `dev`. Vertically complete and independently validated slices are promoted from the development preview to `main`.

The upstream package has no runtime dependency on the development preview. Its public namespace and package paths are exclusively `megatron.experimental.agent_compose`.

## Architecture

The package establishes three layers:

- `primitive`: reusable, independently validated components built from Megatron Core;
- `model`: model-family configuration and composition from validated primitives;
- `runtime`: backend-neutral lifecycle and training orchestration through model protocols.

Dependencies flow from runtime to model to primitive. Model and primitive code may use explicitly shared runtime contracts, but they do not import runtime backends.

The package tree is:

```text
experimental/agent_compose/
  docs/
  megatron/
    experimental/
      agent_compose/
        primitive/
        model/
        runtime/
  skills/
  tests/
```

The public runtime entrypoint is `megatron.experimental.agent_compose.runtime`.

## Runtime Interface

The required Runtime methods form the pretraining tier:

- `build_model`
- `save_checkpoint` / `load_checkpoint`
- `train_mode` / `eval_mode`
- `forward_backward`
- `zero_grad`
- `optimizer_step`
- `lr_scheduler_step`

Implementing `export_weights` enables the `rl_ready` tier. Implementing both `export_weights` and `to` enables the `rl_best` tier.

The skeleton also provides `register_runtime` and `create_runtime`, but intentionally registers no built-in backend. Concrete backends remain separate follow-up PRs.

## Incubation Lifecycle

Incomplete prototypes remain on `dev`. Changes promoted to `main` must be functionally complete and validated for their declared scope. Successful capabilities graduate to Megatron Core, Megatron Bridge, Automodel, or the owning integration project according to ownership. Agent Compose switches to the graduated implementation and demonstrates parity before duplicate incubating code is removed.

## Scope

This PR intentionally does not include:

- concrete primitive implementations;
- concrete model implementations or a model registry;
- built-in runtime backends;
- benchmark or application integrations.

Those surfaces will be proposed in separate, small PRs based on `main`.

## Skills

The initial skills define global Agent Compose constraints, structural skill linting, the primitive review contract, model composition requirements, and end-to-end Runtime validation requirements.

Each future primitive PR should include or update its corresponding leaf skill, reference, and validation path.

## Validation

```text
PYTHONPATH="$PWD:$PWD/experimental/agent_compose" \
  python -m pytest -q experimental/agent_compose/tests/unit

9 passed
```

```text
ruff check experimental/agent_compose/megatron \
  experimental/agent_compose/tests

All checks passed
```

```text
ruff format --check experimental/agent_compose/megatron \
  experimental/agent_compose/tests

14 files already formatted
```

### CODE REVIEW

File: experimental/agent_compose/README.md [hunk 1]
```diff
@@ -1,45 +1,107 @@
...
+Agent Compose is the upstream home for incrementally reviewed Megatron Lite
+components. It makes Megatron-LM development agentic-native by composing
+Megatron Core primitives with coding agents, rather than introducing a new
+standalone product or training stack.
 
-This directory is a placeholder that establishes the location and naming for
-the upstreamed work. Content will land here incrementally as a series of small,
-reviewable PRs.
+`experimental/agent_compose` is the project and review location. The Python
+package keeps the public namespace `megatron.lite`; `experimental.agent_compose`
```
Comment (sbhavani) [inline experimental/agent_compose/README.md]: Could we avoid preserving `megatron.lite` upstream? Since this is intended as an experimental use case, a namespace such as `megatron.experimental.compose` would make its status clearer. You can always alias to something shorter as needed but the canonical Python namespace should reflect the product positioning.

CC @ericharper

File: experimental/agent_compose/README.md [hunk 2]
```diff
@@ -1,45 +1,107 @@
...
+is not an import path.
 
-## Preview
+## Main And Dev
```
Comment (sbhavani) [inline experimental/agent_compose/README.md:12]: Could you add a section or paragraph on Incubation? 

I was thinking something like: "Agent Compose is an experimental incubation surface for Megatron. Successful work should graduate into the appropriate existing areas."

Examples:
RL-facing interfaces → Megatron Core
training loop → Megatron Bridge or Automodel

File: experimental/agent_compose/docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+or distributed runtime lifecycle.
+
+### Runtime
+
+`megatron.lite.runtime` owns the backend-neutral lifecycle: model construction,
+mode changes, forward/backward microbatch orchestration, checkpoint dispatch,
+optimizer and scheduler steps, weight export, and optional device offload.
+Concrete backends implement the public `Runtime` interface.
+
+## Dependency Direction
```
Comment (sbhavani) [inline experimental/agent_compose/docs/architecture.md:31]: Consider a lifecycle section covering how features can be promoted to Megatron Core

**Review by sbhavani — CHANGES_REQUESTED** [review]

File: experimental/agent_compose/README.md (hunk 1 above)
Comment (ISEEKYAN) [reply experimental/agent_compose/README.md]: Thanks, the experimental positioning makes sense.

  One concern is that `compose` is narrower than the intended scope. Agent Compose
  includes a backend-neutral runtime interface, model and primitive boundaries,
  validation contracts, and the agent skills layer; it is not only a model or
  primitive composition utility.

  Would `megatron.experimental.agent_compose` address the positioning concern
  while keeping the canonical namespace aligned with the project name and its
  broader scope?

  The `dev/experimental/lite` preview can continue using `megatron.lite` while
  components are incubated. We would not add a `megatron.lite` alias upstream in
  this PR.

File: experimental/agent_compose/README.md (hunk 2 above)
Comment (ISEEKYAN) [reply experimental/agent_compose/README.md:12]: Agreed on framing Agent Compose as an incubation surface.

  We would like to make one guarantee explicit: incubation here should not mean
  landing partial or non-runnable fragments. Incomplete work remains in the
  `dev/experimental/lite` preview. Changes promoted to `main` should be vertically
  complete and validated for their declared scope.

  When a reusable capability graduates to Megatron Core, Megatron Bridge, or
  Automodel, Agent Compose should first switch to consuming the graduated
  implementation and validate parity. Only then should the duplicate incubating
  implementation be removed. This keeps the Agent Compose path functional rather
  than leaving behind disconnected stubs.

  We will add this distinction to the Incubation section.

File: experimental/agent_compose/docs/architecture.md (hunk 3 above)
Comment (ISEEKYAN) [reply experimental/agent_compose/docs/architecture.md:31]: Agreed. We will add a lifecycle section covering:

  1. prototype and iterate on `dev`;
  2. promote a vertically complete, referenced, and validated slice into Agent
     Compose;
  3. validate correctness, composition, end-to-end behavior, and performance;
  4. identify the long-term owner based on the capability:
     - reusable primitives and backend-neutral interfaces -> Megatron Core;
     - training orchestration -> Megatron Bridge or Automodel;
     - integration-specific behavior -> the owning integration project;
  5. switch Agent Compose to consume the graduated implementation;
  6. remove the duplicate implementation only after parity is demonstrated.

  The goal is for Agent Compose to remain an end-to-end coherent incubation and
  integration surface, not a collection of partial staging artifacts.

File: experimental/agent_compose/README.md (hunk 1 above)
Comment (ericharper) [reply experimental/agent_compose/README.md]: We need to change the name everywhere. `lite` indicates that we are maintaining a fork of megatron which we are not. That goes against the core design principles of megatron.

**Review by sbhavani — APPROVED** [review]
Looks good! It now uses Agent Compose and has the right product positioning
